SingleLinkedList.java

```diff
@@ -0,0 +1,157 @@
+////////////////////////////////////////////////////////////////////////////////
+// Author: Ben Kuschner
+////////////////////////////////////////////////////////////////////////////////
+
+public class SingleLinkedList<T> {
+   
+   /////////////////////////////////////////////////////////
+   // Private link class
+   /////////////////////////////////////////////////////////
+   private class Link<T> {
+      // Member Variables
+      private T data;
+      private Link<T> next;
+
+      // Constructor(s)
+      public Link (T data) {
+         this.data = data;
+         next = null;
+      }
+   
+      // Member methods
+      public T getData() {
+         return this.data;
+      }
+      
+      //deleted hasNext() method because I did not use it
+   
+      public Link<T> getNext() {
+         return next;
+      }
+
+      public void setNext(Link<T> next) {
+         this.next = next;
+      }
+
+   } // end of class link
+
+   /////////////////////////////////////////////////////////
+   // Single Linked List
+   /////////////////////////////////////////////////////////
+
+   // Member variables
+   private Link<T> first;
+   private Link<T> last;
+   private int length;
+   	
+   // Constructor(s)
+   public SingleLinkedList() {
+      first = null;
+      last = null;
+      length = 0;
+   }
+
+   // Member methods
+	
+   public boolean isEmpty () {
+      return first == null;
+   }
+	
+   // Inserts a node at the front of the list.
+   public void insertFront(final T data) {
+      Link<T> newLink = new Link<T>(data);
+      newLink.setNext(first);
+      first = newLink;
+      ++length;
+      if(length == 1) last = first;
+   }
+
+   // Inserts a node at the back of the list
+   public void insertBack(final T data) {
+	  if (!isEmpty()) {
+         last.setNext(new Link<T>(data));
+         last = last.getNext();
+	  }
+	  else {
+		  first = new Link<T>(data);
+		  last = first;
+	  }
+	  ++length;
+   }
+
+   // Returns the amount of objects in the list.
+   public int length() {
+      return length;
+   }
+	
+   // Removes the first node and returns the value stored in it.
+   public T popFront() {
+      T popped = first.getData();
+      first = first.getNext();
+      --length;
+      return popped;
+   }
+
+   // Removes the last node and returns the value stored in it.
+   // O(n)
+   public T popBack() {
```

Comment:
This could probably be improved to O(1) if you stored a reference to the second to last node. However, that may complicate the implementation too much so either way you go for sounds good with this.